### DISCUSSION TITLE
[Proposal] Allow to use initializer instead of constructor at records

### DISCUSSION THREAD

**Discussion by SWATOPLUS** [dotnet/csharplang#6849] [Language Ideas]
We need to allow to use initializer instead of constructor at records.

Before records was introduced, we had used plain C# objects with properties.
```csharp
public class MutDto
{
	public int IntProp { get; set; }
	public string StringProp { get; set; }
}
```
And init them like this:
```csharp
var mutDto = new MutDto
{
	IntProp = 42,
	StringProp = "anytext",
};
```
Now we have init-only and required properties that can make DTO immutable and secure.

```csharp
public class Dto
{
	public required int IntProp { get; init; }
	public required string StringProp { get; init; }
}
```
And it can be initialized in a same way:
```csharp
var dto = new Dto
{
	IntProp = 42,
	StringProp = "anytext",
};
```
We also have records now and it reduces boilerplate code.

```csharp
public record DtoRec(int IntProp, string StringProp);
```
But we still not able to use initializer, we have to use constructor:
```csharp
var dtoRec = DtoRec(42, "anytext");
	
var dtoRec2 = DtoRec(
	IntProp: 42,
	StringProp: "anytext" // no trailing comma support
);
```
But it would be great if we use same code:
```csharp
// this is compile error now
var dtoRecInCSharpNext = new DtoRec
{
	IntProp = 42,
	StringProp = "anytext",
};
```
From developer perspective we can reduce boilerplate in declaration of DTOs changing them from plain object to record without changing code which uses this DTOs. We also get free comparer and object copy implementation.

We really need this feature!

**Comment by Ungerfall** [comment]
Looks convenient. The only issue I see is that an object initializer calls a parameterless ctor. A record doesn't have one.

**Reply by SWATOPLUS** [reply]
If record have no body, mean {}-section, but only primary constructor, compiler can add parameter-less constructor and required properties, otherwise do current behaviour.

**Reply by xtadex** [reply]
At this moment compiler adds only the constructor with parameters. Adding a new parameterless constructor to the class is a breaking change.

**Comment by HaloFour** [comment]
You can declare records without the primary constructor:

```cs
public record Dto
{
	public required int IntProp { get; init; }
	public required string StringProp { get; init; }
}
```

**Reply by bernd5** [reply]
Maybe the properties from the primary constructor could have now `required`, too.
And a default ctor could be emitted (which should have the `Obsolete` / `CompilerFeatureRequired` Attribute, too).

**Reply by HaloFour** [reply]
This could also be accomplished via source generators since the compiler will respect properties that are explicitly defined in code.

**Reply by SWATOPLUS** [reply]
```csharp
public record Dto
{
	public required int IntProp { get; init; }
	public required string StringProp { get; init; }
}
```
Is much longer then, `public record Dto(int IntProp, string StringProp);`

We need to have simple way to define flexible DTO's. Record is designed for data types, it should be use as a short definition of plain data-type with no logic and so on. Records is not popular feature because of differences of default class.

When we tell about records we think about one-line declaration with a primary constructor, but not about multi-line boilerplate.

Primary constructor is optimization of useless case, but short property declaration is case that very very simplifies coding data types, and implement business logic with data transformation.

So, we need some other cases to instantiate plain-record (record with primary constructor only) using initialized. May be by adding some additional attributes like (PlainDataRecordAttribure), or generation parameter-less construction, or even create new type of data structure.
`public dataclass Dto(int IntProp, string StringProp);` and `Dto` will be immutable and `Dto.Builder` - mutable.

But we really need to simple work with DTO's!

**Reply by bernd5** [reply]
@HaloFour good point

But the generated default ctor would need to call the primary ctor (that is a bit ugly):

```csharp
new Foo(){
  I  = 1   
};

record Foo(int I){
    public required int I = I;
    
    public Foo() : this(default(int))
    {}
}
```

**Reply by HaloFour** [reply]
@SWATOPLUS 

There are a _lot_ of design meeting notes discussing records feature over the last 4-5 years.  Ultimately while there is intent to make them simpler DTOs their primary feature is to provide value identity for a type.  The form with a primary constructor is not a simple DTO, it's intended to be used as a case class in pattern matching scenarios and as a likely stepping stone to discriminated unions.  Nominal records are more DTO-like, although I would question the value of value identity in any such type.  There were [discussions](https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-05-27.md#short-property-syntax) to add shorthand for the property declarations to try to make them as concise as positional record declarations, however there doesn't seem to be a lot of impetus to move forward on that.

**Reply by SWATOPLUS** [reply]
@HaloFour So, in this case we need one more construction for DTO's. As I see there is plenty of tech problems to adjust records. 

New construction should be like `public dataclass Dto{int IntProp, string StringProp};` and should be turned to this:

```csharp
public record Dto
{
	public required int IntProp { get; init; }
	public required string StringProp { get; init; }
}
```
